src/uct/cuda/cuda_copy/cuda_copy_ep.c

```diff
@@ -97,40 +98,41 @@ uct_cuda_copy_get_mem_type(uct_md_h md, void *address, size_t length)
     return mem_info.type;
 }
 
-static UCS_F_ALWAYS_INLINE ucs_status_t uct_cuda_copy_get_ctx_rsc(
-        uct_cuda_copy_iface_t *iface, uct_cuda_copy_ctx_rsc_t **ctx_rsc)
+static UCS_F_ALWAYS_INLINE ucs_status_t uct_cuda_copy_ctx_rsc_get(
+        uct_cuda_copy_iface_t *iface, uct_cuda_copy_ctx_rsc_t **ctx_rsc_p)
 {
-    CUcontext current_ctx;
+    CUcontext ctx;
     ucs_status_t status;
+    khiter_t iter;
+#if CUDA_VERSION >= 12000
+    unsigned long long UCS_V_UNUSED ctx_id;
+    CUresult UCS_V_UNUSED result;
+#endif
 
-    status = UCT_CUDADRV_FUNC_LOG_ERR(cuCtxGetCurrent(&current_ctx));
-    if (status != UCS_OK) {
+    status = UCT_CUDADRV_FUNC_LOG_ERR(cuCtxGetCurrent(&ctx));
+    if (ucs_unlikely(status != UCS_OK)) {
         return status;
-    } else if (current_ctx == NULL) {
+    } else if (ucs_unlikely(ctx == NULL)) {
         ucs_error("no context bound to calling thread");
         return UCS_ERR_IO_ERROR;
     }
 
-    return uct_cuda_copy_iface_get_ctx_rsc(iface, current_ctx, ctx_rsc);
-}
-
-static UCS_F_ALWAYS_INLINE ucs_status_t
-uct_cuda_copy_get_short_stream(uct_cuda_copy_iface_t *iface, CUstream *stream)
-{
-    uct_cuda_copy_ctx_rsc_t *ctx_rsc;
-    ucs_status_t status;
-
-    status = uct_cuda_copy_get_ctx_rsc(iface, &ctx_rsc);
-    if (status != UCS_OK) {
-        return status;
+    iter = kh_get(cuda_copy_ctx_rscs, &iface->ctx_rscs, (khint64_t)ctx);
+    if (ucs_unlikely(iter == kh_end(&iface->ctx_rscs))) {
+        return uct_cuda_copy_ctx_rsc_create(iface, ctx, ctx_rsc_p);
     }
 
-    status = uct_cuda_copy_init_stream(&ctx_rsc->short_stream);
-    if (status != UCS_OK) {
-        return status;
-    }
+    *ctx_rsc_p = &kh_value(&iface->ctx_rscs, iter);
+
+#if CUDA_VERSION >= 12000
+    ucs_assertv((result = cuCtxGetId(ctx, &ctx_id)) == CUDA_SUCCESS,
+                "cuCtxGetId failed %s",
+                uct_cuda_base_cu_get_error_string(result));
+    ucs_assertv(ctx_id == (*ctx_rsc_p)->ctx_id,
+                "context id %llu does not match the cached value %llu", ctx_id,
+                (*ctx_rsc_p)->ctx_id);
+#endif
```

Comment:
if context id is changed, drop the hash value and call uct_cuda_copy_ctx_rsc_create